src/ca/mcgill/cs/jetuml/gui/DiagramCanvasController.java

```diff
@@ -320,6 +326,34 @@ else if(!aSelectionModel.contains(element.get()))
 		}
 	}
 
+	/* Recursively reorder the node to be on top of its parent's all children. 
+	 * If the node is not a child node or it does not have a parent, check if 
+	 * the node is a root node of the diagram and reorder it to be on the top 
+	 * of other root nodes.
+	 */
+	private void reorderSelectedNode(Node pNode) 
+	{
+		assert pNode != null;
+		ParentNode parent = null;
+		if (pNode instanceof ChildNode)
+		{
+			parent = ((ChildNode)pNode).getParent();
+		}
+		if(parent != null) 
+		{
+			// Move the child node to the top of all other children  
+			parent.removeChild((ChildNode)pNode);

```

Comment:
Maybe encapsulate this operation in a method `moveChildToLastPlace(ChildNode)` within `ParentNode`, which will also facilitate testing.

Reply:
If I add the method to the interface of the _ParentNode_, I'll need to implement the method in the three subclasses that implement the _ParentNode_, which causes code redundancy since the method is doing the same thing for these subclasses. Adding an abstract ParentNode class could solve this concern but I think this might be unnecessary for this two-line operation. 

Maybe we could add the method as a default method with implementation in the interface of _ParentNode_, but I'm not sure if this follows the coding convention of JetUML since other interfaces only specify method signatures.

Reply:
Also, I wonder if I only need to test for **placeOnTop(Node): void** and **containsAsRoot(Node):boolean** in _Diagram_ class; since in _DiagramCanvasController_ , I only add 4 lines of code in the method  _handleSelection(MouseEvent): void_.

Reply:
Let's try defining `moveChildToLastPlace(ChildNode)` as a default method in `ParentNode`. As for `placeOnTop` and `containsAsRoot(Node)` I think it's indeed worth it to move these to `Diagram`, because conceptually they are really operations on the diagram. Ideally I would like to reduce to zero the number of code locations than modify the state of the `Diagram` outside of the `Diagram` class.